### PR TITLE
Orphans improvements part 2

### PR DESCRIPTION
**Pull request by michaelsutton**

### CODE REVIEW

File: protocol/flows/src/flowcontext/orphans.rs [hunk 1]
```diff
@@ -63,11 +78,45 @@ impl OrphanBlocksPool {
...
-            debug!("Orphan blocks pool size exceeded. Evicting a random orphan block.");
-            // Evict a random orphan in order to keep pool size under the limit
-            if let Some((evicted, _)) = self.orphans.swap_remove_index(rand::thread_rng().gen_range(0..self.max_orphans)) {
-                debug!("Evicted {} from the orphan blocks pool", evicted);
+            let mut eviction_succeeded = false;
+            debug!("Orphan blocks pool size exceeded. Trying to evict a random orphan block.");
+            // Retry up to a logarithmic number of times
+            for i in 0..self.max_orphans_log {
+                // Evict a random orphan in order to keep pool size under the limit
+                let rand_index = rand::thread_rng().gen_range(0..orphan_ancestors.len());
```
Comment (someone235) [inline protocol/flows/src/flowcontext/orphans.rs]: Switch to orphan pool size

File: protocol/flows/src/flowcontext/orphans.rs [hunk 2]
```diff
@@ -90,19 +139,24 @@ impl OrphanBlocksPool {
...
+        } else {
+            OrphanOutput::Unknown
         }
-        self.get_orphan_roots(consensus, orphan).await
     }
 
-    /// Internal get roots method. Assumes 'orphan' is within the pool
-    async fn get_orphan_roots(&self, consensus: &ConsensusProxy, orphan: Hash) -> OrphanOutput {
+    /// Internal get roots method
+    async fn get_orphan_roots(&self, consensus: &ConsensusProxy, mut queue: VecDeque<Hash>) -> FindRootsOutput {
```
Comment (someone235) [inline protocol/flows/src/flowcontext/orphans.rs:153]: Some explanation about what is `queue` would be nice

File: components/consensusmanager/src/batch.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+use kaspa_consensus_core::{api::BlockValidationFuture, block::Block};
+use std::fmt::Debug;
+
+pub struct BlockProcessingBatch {
```
Comment (someone235) [inline components/consensusmanager/src/batch.rs:4]: I think it's more readable if you wrap the whole thing with an `Option`, instead of internal fields?

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — APPROVED** [review]

File: components/consensusmanager/src/batch.rs (hunk 3 above)
Comment (michaelsutton) [reply components/consensusmanager/src/batch.rs:4]: Sometimes only one of the task vecs is taken and the others remain for future processing

**Pull request merged**